Civi/Share/Message.php

```diff
@@ -23,15 +23,69 @@
  */
 class Message {
 
-  /** @var array list of change ids */
-  protected array $change_ids = [];
+  /**
+   * @phpstan-var list<\Civi\Share\Change>
+   */
+  protected array $changes = [];
 
   /**
    * Create a new, empty message
    */
-  public function __construct()
-  {
+  public function __construct() {
+
+  }
+
+  public static function createFromSerializedMessage(array $serializedMessage, ?\DateTime $receivedDate = NULL) {
+    $message = new self();
+
+    try {
+      $sourceNode = ShareNode::get()
+        ->addSelect('id')
+        ->addWhere('short_name', '=', $serializedMessage['payload']['sender'])
+        ->execute()
+        ->single()['id'];
+    }
+    catch (\CRM_Core_Exception $e) {
+      throw new \RuntimeException(
+        sprintf(
+          'Could not identify source node %s for change message %s.',
+          $serializedMessage[' payload']['sender'],
+          $serializedMessage['id']
+        ),
+        0,
+        $e
+      );
+    }
 
+    // TODO: Add metadata properties.
+
+    // Add changes.
+    if (!is_array($serializedMessage['payload']['changes'])) {
+      throw new \RuntimeException(
+        sprintf('Could not parse changes for change message %s.', $serializedMessage['id'])
+      );
+    }
+    try {
+      foreach ($serializedMessage['payload']['changes'] as $serializedChange) {
+        $message->addChange(Change::createFromSerialized($serializedChange, $sourceNodeId, $receivedDate));
+      }
+    }
+    catch (\RuntimeException $e) {
+      throw new \RuntimeException(
+        sprintf('Could not parse changes for change message %s.', $serializedMessage['id'])
+      );
+    }
+
+    return $message;
+  }
+
+  public function getChangeIds(bool $onlyPersisted = FALSE): Iterable {
```

Comment:
Boolean parameters are always a bit problematic because there meaning is unclear when the method signature is not visible. So it might make sense to have a method `getPersistedChangeIds()` instead.